### DISCUSSION TITLE
Precision control of references in NIEM 6

### DISCUSSION THREAD

**Discussion by iamdrscott** [niemopen/ntac-admin#69] [General]
Message designers do not have much control over references in NIEM 5.  Every element is referenceable by default.  The designer can subset *structures.xsd* to remove the `ref` and `uri` attributes, and then no element is referenceable. That's it.

Message designers have much more control in NIEM 6.  By default, elements with children or attributes are  referenceable.  By default, elements with simple content and no attributes are *not* referenceable.  But you can override the default for all elements of a type, or for particular elements.  In CMF you do this with `cmf:ReferenceableIndicator`; in XSD with `appinfo:referenceableIndicator`.  This works in NIEM 6 because complex types in a message schema do not have to extend `structures:ObjectType`.  For example, if you want to permit references to Person objects but not Education objects, then your message schema might look like this:

```
<xs:complexType name="EducationType">
  <xs:sequence>
    <xs:element ref="nc:EducationDescriptionText" minOccurs="0" maxOccurs="unbounded"/>
    <xs:element ref="nc:EducationAugmentationPoint" minOccurs="0" maxOccurs="unbounded"/>
  </xs:sequence>
</xs:complexType>
  
<xs:complexType name="PersonType">
  <xs:sequence>
    <xs:element ref="nc:PersonEducation" minOccurs="0" maxOccurs="unbounded"/>
    <xs:element ref="nc:PersonAugmentationPoint" minOccurs="0" maxOccurs="unbounded"/>
  </xs:sequence>
  <xs:attribute ref="appliesToParent"/>
  <xs:attribute ref="structures:id"/>
  <xs:attribute ref="structures:ref"/>
</xs:complexType>
```

Observe that in this schema, references to Person objects must use IDREFs; therefore, when you see a Person reference, you are guaranteed that the data is in the message.  There is no such guarantee when URI references are allowed. 

We don't yet have that level of control in CMF.  So I want to either:

1. Replace `cmf:ReferenceableIndicator` with `cmf:ReferenceCode`, with four code values:  none, ref, uri, any.
2. Add a new `cmf:IdentifiableIndicator` property

BTW, this isn't hypothetical.  I'm working with message designers who want more control over references.

Thoughts?

**Comment by cabralje** [comment]
I agree that it would be good to add support to CMF to match the level of control we support in XSD at least.

I was wondering if using a code list would add unnecessary complexity to the CMF.  However, I see other examples of code lists in CMF already, including cmf:NamespaceKindCode and cmf:WhiteSpaceValueCode. We also very well may find other features in CMF that would be better modeled by a code list.

So, I support replacing cmf:ReferenceableIndicator with cmf:ReferenceCode.


[cid:8f10b7c9-9ac2-4dca-91ad-57715dac04b1]

Jim Cabral

Vice President, Court Relations

502-640-4970


Visit the site<https://www.infotrack.com/?utm_medium=email&utm_source=outlook&utm_campaign=signature> | Check out the blog<https://www.infotrack.com/blog/?utm_medium=email&utm_source=outlook&utm_campaign=signature> | Contact us<https://www.infotrack.com/contact-us?utm_medium=email&utm_source=outlook&utm_campaign=signature>



***@***.***?anonymous&ep=signature>            Book time to meet with ***@***.***?anonymous&ep=signature>
________________________________
From: Scott Renner ***@***.***>
Sent: Wednesday, January 10, 2024 2:54 PM
To: niemopen/ntac-admin ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [niemopen/ntac-admin] Precision control of references in NIEM 6 (Discussion #69)


Message designers do not have much control over references in NIEM 5. Every element is referenceable by default. The designer can subset structures.xsd to remove the ref and uri attributes, and then no element is referenceable. That's it.

Message designers have much more control in NIEM 6. By default, elements with children or attributes are referenceable. By default, elements with simple content and no attributes are not referenceable. But you can override the default for all elements of a type, or for particular elements. In CMF you do this with cmf:ReferenceableIndicator; in XSD with appinfo:referenceableIndicator. This works in NIEM 6 because complex types in a message schema do not have to extend structures:ObjectType. For example, if you want to permit references to Person objects but not Education objects, then your message schema might look like this:

<xs:complexType name="EducationType">
  <xs:sequence>
    <xs:element ref="nc:EducationDescriptionText" minOccurs="0" maxOccurs="unbounded"/>
    <xs:element ref="nc:EducationAugmentationPoint" minOccurs="0" maxOccurs="unbounded"/>
  </xs:sequence>
</xs:complexType>

<xs:complexType name="PersonType">
  <xs:sequence>
    <xs:element ref="nc:PersonEducation" minOccurs="0" maxOccurs="unbounded"/>
    <xs:element ref="nc:PersonAugmentationPoint" minOccurs="0" maxOccurs="unbounded"/>
  </xs:sequence>
  <xs:attribute ref="appliesToParent"/>
  <xs:attribute ref="structures:id"/>
  <xs:attribute ref="structures:ref"/>
</xs:complexType>


Observe that in this schema, references to Person objects must use IDREFs; therefore, when you see a Person reference, you are guaranteed that the data is in the message. There is no such guarantee when URI references are allowed.

We don't yet have that level of control in CMF. So I want to either:

  1.  Replace cmf:ReferenceableIndicator with cmf:ReferenceCode, with four code values: none, ref, uri, any.
  2.  Add a new cmf:IdentifiableIndicator property

BTW, this isn't hypothetical. I'm working with message designers who want more control over references.

Thoughts?

**Comment by cdmgtri** [comment]
CMF currently defines attribute `ReferenceableIndicator` on `PropertyType`, but the example above shows the `structures` attributes included on type `PersonType` and excluded on type `EducationType`.

**Customizing `structures` attributes on types**

If you just want to have a single definition of `PersonType`, then perhaps `ReferenceableIndicator` and `IdentifiableIndicator`, or `ReferenceCode`, should appear on CMF class types instead of properties?

**Customizing `structures` attributes on properties**

If you want to be able to say that a Judge has a `ref` and `nc:Person` has `id`, then it makes sense to have this kind of information on properties.  But then you would need to create custom types so you could have different profiles.  The judge property would no longer extend `nc:PersonType` but would either have a new or an anonymous flattened type with custom `structures` attributes.  And the type for `nc:Person` would be defined similarly.

This gives you finer-grain control over references.  Instead of types either having `id` and `ref` together or not at all, it would let you have one property of a type just have an `id` and other properties of that same type  just have a `ref` in your message model.

**Custom property profiles**

You could extend that solution more generally.  You could have a judge and a subject in a message.  You might just want name for the judge, but you've added name, birthdate, SSN, hair color, eye color, etc., to `nc:PersonType` for the subject.

If you switch from the CMF `ReferenceableIndicator` + `IdentifiableIndicator` (or `ReferenceCode`) to include or exclude lists, you could extend the precision control of references to entire profiles for objects.  The judge property could have an include list of just `nc:PersonName` and `structures:uri` (for example purposes).  If the subject property didn't have an include or exclude list, it would get the usually-generated subset.  Or it could have an exclude list of `structures:uri` to get a custom-built type with everything but that.

Note: Requires `structures` attributes and types to be added to the CMF model to allow them to be treated similarly to other properties and types.

**Reply by iamdrscott** [reply]
Good catch.  (Glad you're back!) I think it should appear on both, with properties overriding classes.  

I think we want all of the cool things you describe.  You can hand-jam them right now in a NIEM 6 message schema -- there's nothing to keep you from custom local type declarations.  But I can't generate that XSD from CMFTool now.  I'll get around to it as time permits.  Right now I don't even have a way to specify your custom property profiles in CMF.  Should we add the hooks into CMF now, or have a CMF 6.1 later?
